documentation/projects/proposals/ingestion_server_removal/20240328-implementation_plan_ingestion_server_removal.md

```diff
@@ -0,0 +1,610 @@
+  - POST to each worker's `reindexing_task` endpoint the `start_index` and
+    `end_index` it should handle
+  - Use a Sensor to ping the worker's `task/{task_id}` endpoint until the task
+    is complete, logging the progress as it goes
+- Use
+  [`terminate_instance_in_auto_scaling_group`](https://boto3.amazonaws.com/v1/documentation/api/1.26.86/reference/services/autoscaling/client/terminate_instance_in_auto_scaling_group.html)
+  to terminate the instance. Make sure to set `ShouldDecrementDesiredCapacity`
+  to `True` to ensure that the ASG does not try to replace the instance. This
+  task should use the
+  [`NONE_SKIPPED` TriggerRule](https://airflow.apache.org/docs/apache-airflow/stable/core-concepts/dags.html#trigger-rules)
+  to ensure that the instances are terminated, even if there are upstream
+  failures. (Skips in local env.)
+- Finally, after all tasks have finished (regardless of success/failure), we
+  should have a cleanup task that calls `set_desired_capacity` to 0. Generally
+  this should be a no-op, but if an instance crashes during reindexing (rather
+  than simply failing during reindexing) the ASG will spin up a replacement and
+  Airflow will not automatically clean it up. This task ensures that any
+  dangling instances are terminated.
+
+```{note}
+It is not possible to retry a single indexer worker with this set up, because once a worker fails the instance is actually terminated (rather than simply stopped). If a task that triggers a reindex is cleared after an instance has been terminated, it will simply fail. The entire reindex must be restarted from the first step in this task group.
+
+However, there is a valuable tradoff to this approach: it ensures that all of the indexer workers in a data refresh are identical, while still allowing us to avoid manual deployments every time the indexer logic changes. For example, imagine some changes to the reindexing logic are merged to `main` while a data refresh is actively underway, and a new Docker image is published. If one indexer worker failed, and it were possible to retry **just** that indexer worker, it would use the new Docker image -- leading to inconsistency in the behavior of different workers within a single data refresh.
+```
+
+### Create the Terraform and Ansible resources needed to deploy the new indexer workers
+
+In this step we will add the resources needed to actually configure the ASGs.
+
+Some important notes:
+
+- Currently, the staging and production workers are split into separate
+  environments (i.e., a staging deploy is used to deploy the 2 staging workers
+  separately from the 6 production workers). It is more accurate to view all 8
+  workers as production instances (i.e., part of the production catalog
+  deployment), which merely _operate_ on different environments. As such all 8
+  should be part of the production deployment, but two separate ASGs which are
+  tagged indicating their intended environment.
```

Comment:
I don't really follow this logic of putting both in the production root module.

The staging workers should use staging code, staging secrets, and reference other staging resources. The production workers should do that for production.

Why break the pattern for Airflow? I think the distinction between what is a staging and what is a production resource is bit confused here. Airflow is somewhat external to our deployment environment concept, but insofar as the ingestion workers should use code versioned to match e.g., the API's index mapping expectations, shouldn't they align with the application-level they refer to?

Following this logic would kind of make both Elasticsearch instances go into the production root module. Airflow is kind of like Kibana, which also references a staging resource, but exists in the production root module.

Airflow and Kibana are external to concepts of staged deployments, and cross things over in ways that make it no longer make sense to speak of whether they are "staging" or "production". It could be helpful to get rid of the idea of a deployment "stage" name for services like this that don't have staged deployments.

The indexer workers _do_ have staged deployments, as far as I know, and if they don't then we don't need two separate ASGs, we can just use a single ASG with a max capacity capable of handling both staging and production index at the same time, rather than having separate ASGs :shrug:

Reply:
The thought behind this is that there is not intended to ever be any difference between the staging vs prod indexer workers, except the connection variables; to test updates we don't deploy staging code for the indexer workers, we run the data refresh against the staging environment. (This is basically also true today: we have not historically made use of the staging ingestion server, so the addition of staging_data_refresh DAGs in this project is actually a huge improvement!)

The advantage of having two ASGs is just that it makes it very simple to allow staging and production data refreshes to run simultaneously, without worrying about them interfering with each other. For example, one of the edge cases I raised in the IP is if an instance crashes during reindexing and the ASG replaces it with a new instance; we add a cleanup task that sets the desired capacity to 0 after reindexing is complete to ensure there cannot possibly be any leftover instances. But we cannot do this if the ASG might also be supporting a staging data refresh simultaneously.

Since we don't expect to have to deploy these often I don't think it's unreasonable to require them to be deployed separately, but it also doesn't seem advantageous to me. I wouldn't be strongly opposed to separating the deployments out if that's preferred, though.

Reply:
Gotcha, that's a good way of thinking about it. I wonder if we can think of them as "pools" rather than staging/production. In that sense, I think it's good to keep "environment" as defined in the terraform code to production for both, and indicate the purpose in the service name, "staging-indexer-worker-pool", "production-indexer-worker-pool" (or whatever you think makes sense, just indicating the purpose there rather than by changing the environment label). I realise now that could be what you meant by "tagged indicating their environment" but I interpreted it as changing the "environment" tag, which is a semantically important tag in Ansible and in how our infra resources are organised in relation to other supporting resources.

It might seem like just a nit-pick, but the production environment label relates to things like the load balancer selected, and I'd be massively confused if I saw infra stuff with environment labelled "staging" attached to the production internal load balancer (in a hypothetical future where we used that approach with the sticky cookies I mentioned, or just to make it easier to reference the instances overall).

Reply:
Might need to take a step back here and defer the decision either way. It'll depend on what is easiest to manage within our Ansible playbooks, and whether we'd use separate AMIs for staging and production indexer workers or if the target environment is a runtime variable :thinking:

Reply:
>  I realise now that could be what you meant by "tagged indicating their environment" but I interpreted it as changing the "environment" tag, which is a semantically important tag in Ansible and in how our infra resources are organised in relation to other supporting resources.

That is what I meant, but I didn't realize I was overloading terminology -- thank you for pointing that out, I will absolutely clarify that.

> It might seem like just a nit-pick

Not at all, it's an important distinction and I was also worried that it might be confusing. I really love the '-pool' suggestion for the tags, that's a small difference that makes it much clearer!